File: Python-PKCS11_Samples/sign_using_rsa_pss.py
#!/usr/bin/env python3
#*********************************************************************************
#                                                                                *
# This file is part of the "luna-samples" project.                               *
#                                                                                *
# The "luna-samples" project is provided under the MIT license (see the          *
# following Web site for further details: https://mit-license.org/ ).            *
#                                                                                *
# Copyright © 2025 Thales Group                                                  *
#                                                                                *
#*********************************************************************************

# OBJECTIVE:
# - This sample code demonstrates how to generate RSA keypair and use it for signing.
# - It uses rsa-pss mechanism for signing.


import sys
import os
import getpass
import pkcs11
from pkcs11 import KeyType, Mechanism, MGF
from pkcs11.exceptions import NoSuchKey, PinIncorrect, NoSuchToken

print("\nsign_using_rsa_pss.py\n")


# Prints the syntax for executing this code.
if len(sys.argv)!=2:
	print ("Usage:")
	print ("./sign_using_rsa_pss.py <slot_label>")
	print ("\nExample:")
	print ("./sign_using_rsa_pss.py SP_SKS_SEHSM3\n")
	quit()

slot_label = sys.argv[1]


# Reads P11_LIB environment variable.
try:
	pkcs11_library = os.environ['P11_LIB']
except:
	print("*** P11_LIB environment variable not set. ***")
	print("> export P11_LIB=/usr/safenet/lunaclient/lib/libCryptoki2_64.so\n")
	quit()



co_pass = getpass.getpass(prompt="Crypto officer password: ")
plaintext = input("Enter plaintext to sign : ")
if (len(plaintext)>245):
	print("Plaintext too long.\n")
	quit()


try:
	p11 = pkcs11.lib(pkcs11_library) # Loads pkcs11 library.
	print ("PKCS11 library found at : ", pkcs11_library)

	p11token = p11.get_token(token_label=slot_label) # Find the specified slot.
	print("Token found : ", slot_label)

	with p11token.open(user_pin=co_pass) as p11session: #Opens a new session and logs in as crypto officer.
		print("Login success.")
		rsa_pub, rsa_pri = p11session.generate_keypair(pkcs11.KeyType.RSA, 2048, store=False) # Generates rsa-2048 keypair as session objects.

		print ("RSA-2048 keypair generated.")
		signature = rsa_pri.sign(plaintext, mechanism=Mechanism.RSA_PKCS_PSS) # Generates signature using rsa-pss-mgf-256 mechanism.
		print ("Plaintext signed.")
		if(rsa_pub.verify(plaintext, signature, mechanism=Mechanism.RSA_PKCS_PSS)):
			print ("Signature verified.\n")
		else:
			print ("Signature verification failed.\n");

		print ("Plain text	: ", plaintext)
		print ("Plain text (hex): ", plaintext.encode().hex())
		print ("Signature	: ", signature.hex())
		print ()

except PinIncorrect:
	print ("Incorrect crypto officer pin.\n")
except NoSuchToken:
	print ("Incorrect token label.\n")
except:
	print (sys.exc_info()[0])
	print ()

File: Node-PKCS11_Samples/sign_using_rsa_pss.js
#!/usr/bin/env node
/*********************************************************************************
 *                                                                                *
 * Ported from ThalesGroup/luna-samples. The original project is provided under   *
 * the MIT license (https://mit-license.org/).                                    *
 *                                                                                *
 * Copyright © 2025 Thales Group (original samples)                               *
 *                                                                                *
 *********************************************************************************

 * OBJECTIVE:
 * - This sample code demonstrates how to generate an RSA keypair and use it for signing.
 * - It uses RSA-PSS mechanism for signing.
 */

"use strict";
const { graphene, withSession, usageAndExit, getPlaintext, toHex } = require("./lib/helper");

console.log("\nsign_using_rsa_pss.js\n");

if (process.argv.length !== 3) {
  usageAndExit([
    "Usage:",
    "node sign_using_rsa_pss.js <slot_label>",
    "",
    "Example:",
    "node sign_using_rsa_pss.js myPartition\n",
  ]);
}

const slotLabel = process.argv[2];

(async () => {
  const plaintext = await getPlaintext("Enter plaintext to sign : ", 245);
  await withSession(slotLabel, async (session) => {
    const keys = session.generateKeyPair(
      graphene.KeyGenMechanism.RSA,
      {
        keyType: graphene.KeyType.RSA,
        modulusBits: 2048,
        publicExponent: Buffer.from([0x01, 0x00, 0x01]),
        token: false,
        verify: true,
      },
      {
        keyType: graphene.KeyType.RSA,
        token: false,
        private: true,
        sign: true,
      }
    );
    console.log("RSA-2048 keypair generated.");

    // Hash outside then sign with RSA_PKCS_PSS (match python-pkcs11 RSA_PKCS_PSS usage)
    const digest = session.createDigest(graphene.MechanismEnum.SHA256);
    const hash = digest.once(Buffer.from(plaintext, "utf8"));

    const alg = {
      name: graphene.MechanismEnum.RSA_PKCS_PSS,
      params: new graphene.RsaPssParams(
        graphene.MechanismEnum.SHA256,
        graphene.RsaMgf.MGF1_SHA256,
        32
      ),
    };

    const sign = session.createSign(alg, keys.privateKey);
    const signature = sign.once(hash);
    console.log("Plaintext signed.");

    const verify = session.createVerify(alg, keys.publicKey);
    const ok = verify.once(hash, signature);
    console.log(ok ? "Signature verified.\n" : "Signature verification failed.\n");

    console.log("Plain text\t: ", plaintext);
    console.log("Plain text (hex): ", Buffer.from(plaintext, "utf8").toString("hex"));
    console.log("Signature\t: ", toHex(signature));
    console.log();
  });
})();

